### PR TITLE
Make type_expr private

### PR DESCRIPTION
**Pull request by garrigue**
This PR makes `Types.type_expr` a private record, so that one cannot directly create or modify it.
The public definition is in submodule `Internal`, with conversion (identity) functions.

To make this definition easier to use, `Btype.mark_type_node` is extended with two new optional arguments, `guard` and `after`, which are called with the `repr`ed type, respectively before marking it (returning whether to mark it or not) and after marking it.

We have plans for further abstraction, but this seems a good first step.

@t6s co-authored this PR.

### CODE REVIEW

**Comment by gasche** [comment]
This looks interesting, thanks! (It's also joint work with @t6s, one of the first such contribution: welcome!)

I have the impression that `mark_type_node` is almost always passed an `after` parameter. To me this suggests that the API could be

```ocaml
val mark_type_node: type_expr -> ?guard:(type_expr -> bool) -> (type_expr -> unit) -> unit
```

to be used as

```ocaml
mark_type_node ty @@ fun ty ->
...
```

(or with just `ignore` in the cases that currently use the default `after` value; but I think that they could all be rewritten to make meaningful use of their parameter.)

**Comment by garrigue** [comment]
I have no strong opinion as to whether `after` should be optional or not, but do you really think that this `@@` syntax is going to be more readable?
It's starting to look Haskellish.

**Comment by gasche** [comment]
Don't use `@@` if you don't like it, but I think `after` should be mandatory.

**Comment by trefis** [comment]
On a not-exactly-related-but-still-somewhat-relevant note: is there some reason to keep using the types marking mechanism instead of a table indexed by the `type expr` id (as is done in `Ctype.lower_contravariant` for instance)?

**Comment by t6s** [comment]
Ctype.closed_class is the only place where mark_type_node is called without `after`, but its code is complicated by `try .. with`. Can we fit this logic into a use of `after`?

**Comment by gasche** [comment]
I would try


```diff
let closed_class params sign =
  let ty = object_fields (repr sign.csig_self) in
  let (fields, rest) = flatten_fields ty in
  List.iter mark_type params;
  mark_type rest;
  List.iter
    (fun (lab, _, ty) -> if lab = dummy_method then mark_type ty)
    fields;
  try
-    mark_type_node (repr sign.csig_self);
+    mark_type_node (repr sign.csig_self) @@ fun csig_self ->
    List.iter
      (fun (lab, kind, ty) ->
        if field_kind_repr kind = Fpresent then
        try closed_type ty with Non_closed (ty0, real) ->
          raise (CCFailure (CC_Method (ty0, real, lab, ty))))
      fields;
-    mark_type_params (repr sign.csig_self);
+    mark_type_params csig_self;
    List.iter unmark_type params;
    unmark_class_signature sign;
    None
  with CCFailure reason ->
    mark_type_params (repr sign.csig_self);
    List.iter unmark_type params;
    unmark_class_signature sign;
    Some reason
```

**Comment by gasche** [comment]
We could also factorize as

```ocaml
let closed_class params sign =
  let ty = object_fields (repr sign.csig_self) in
  let (fields, rest) = flatten_fields ty in
  List.iter mark_type params;
  mark_type rest;
  List.iter
    (fun (lab, _, ty) -> if lab = dummy_method then mark_type ty)
    fields;
  mark_type_node (repr sign.csig_self) @@ fun csig_self ->
  let reason =
    try
      List.iter
      (fun (lab, kind, ty) ->
        if field_kind_repr kind = Fpresent then
        try closed_type ty with Non_closed (ty0, real) ->
          raise (CCFailure (CC_Method (ty0, real, lab, ty))))
      fields;
      None
    with CCFailure reason ->
      Some reason in
  mark_type_params csig_self;
  List.iter unmark_type params;
  unmark_class_signature sign;
  reason
```

**Comment by gasche** [comment]
Note: we would not need an exception at all if we had my baby `List.map_option` from #9630.

**Comment by garrigue** [comment]
@gasche It looks like you misunderstood the role of `after`.
It is only called when the node has to be marked (i.e. `level >= lowest_level`).
So your suggested changes for `closed_class` would be incorrect.

My temporary conclusion is that the current proposed API of using an optional argument is better, because it does not invite people to always add some post-processing, which could be wrongly scoped.
Same thing for `@@`, which changes the execution path without parentheses (have we learnt nothing from the `if then` debacle :-)

As for @trefis 's comment, the use of a hash table in `lower_contravariant` comes from the possibility of expansion, which would invalidate the invariants of marking. But I agree that `mark_type` goes against the goal of abstracting the type representation (and requires exclusive access to the type graph, which is hard to enforce statically), and may eventually have to be replaced by something else. Not in this PR, though.

**Comment by gasche** [comment]
Good catch; in this case `mark_type_node (repr sign.csig_self) ignore` would be fine?

**Comment by gasche** [comment]
> My temporary conclusion is that the current proposed API of using an optional argument is better, because it does not invite people to always add some post-processing, which could be wrongly scoped.

After thinking about this, I think that either APIs are just as error-prone as each other, and I am still in favor of a mandatory and non-labelled `after` parameter. If the error in my newcomer suggestion above can be attributed to something else than carelessness, it would dig in the following direction: why is this function using `mark_type_node` in such a different way than all other callsites? Should it not be using another marking function?

**Comment by garrigue** [comment]
So you want use to re-split it in `mark_type_node` (as before) and `mark_type_node_with` ?
This makes all the other calls longer, but if you think  so...

Note that this call is not that different: `fields` being a list of subnodes of `sign.csig_self`, the `List.iter` just after could be inside `after`. But since we do not refer directly to `csig_self` in it, there is little incentive to do that.

I think the real choice is whether it is ok to use optional arguments to factorize APIs.
This is certainly the case in libraries, but I understand if there is some resistance inside the compiler.

**Comment by t6s** [comment]
The last commit makes `after` mandatory and separates the particular use-case of `mark_type_node` in `closed_class` into a new function `mark_type_node_only`.  Do you think this makes the changes look better? @gasche

File: typing/btype.ml
Comment (lpw25) [inline typing/btype.ml]: If this isn't used, why are we leaving it in a comment?

File: typing/btype.ml
Comment (lpw25) [inline typing/btype.ml]: There are a few lines like this. Maybe add an `Internal.set_desc` to make them easier to read.

File: typing/btype.ml
Comment (lpw25) [inline typing/btype.ml]: And an `Internal.set_level` to use here too.

File: typing/types.mli
Comment (lpw25) [inline typing/types.mli:241]: Rather than exposing the actual type and conversions, why not expose the three setters and a constructor function? Those are all the operations that you are actually exposing, and it seems better to just use them directly.

File: typing/datarepr.ml
Comment (lpw25) [inline typing/datarepr.ml]: Maybe move this into `Btype` with a name like `dummy_type`.

File: typing/btype.mli
Comment (lpw25) [inline typing/btype.mli]: I'm not a fan of this interface. Why not just use the much simpler:
```ocaml
val mark_type_node : type_expr -> bool
```
that returns `false` if the node was already marked and returns `true` and marks the node if was not already marked.

**Review by lpw25 — COMMENTED** [review]
This is a good idea. I think the code could be cleaner though, so I've left some comments.

File: typing/btype.mli
Comment (garrigue) [reply typing/btype.mli]: Good idea. We should try this.

File: typing/btype.ml
Comment (garrigue) [reply typing/btype.ml]: The idea here is that one should not directly modify types, so providing a function that does it in `types.mli` seems a mixed message. Or change the name `Internal` to something more scary like `Unsafe_access`.

File: typing/types.mli
Comment (garrigue) [reply typing/types.mli:241]: Explained above. This is not an abstract API, but rather a gaping hole to allow `Btype` to do its job.
Ideally all the functions of `Btype` would be in `Types`, and this wouldn't be needed.

File: typing/btype.mli
Comment (garrigue) [reply typing/btype.mli]: I tried on a few cases.
While it is an improvement on the original style, it is still more verbose than the functional version.
Typically, something like
```ocaml
mark_type_node ty ~after:(it.it_do_type_expr it)
```
must be rewritten as
```ocaml
let ty = repr ty in
if mark_type_node ty then it.it_do_type_expr it ty
```
which lacks focus.
Since the marking function has to manipulate the low level representation of types, our idea was to make the code more regular, and I think the `mark_type_node ty ~after:` is successful for that. @gasche 's remark is actually a testament to that: it shows that the code is so regular that the version without `after` was only used once.

File: typing/types.mli
Comment (garrigue) [reply typing/types.mli:241]: By the way, we are working also on the next step, where `type_expr` would be abstract rather than private, and constructor and access functions would be provided. But this is not in the scope of this PR, which only restricts mutation.

Comment (lpw25) [reply typing/types.mli:241]: I appreciate that it is a backdoor rather than the intended interface -- and a scarier name would be great too -- but that doesn't mean that the backdoor shouldn't itself have a well-designed interface. As it is now, the code in Btypes is being made harder to read than it needs to be.

File: typing/btype.mli
Comment (lpw25) [reply typing/btype.mli]: I think that the second version you provide is noticeably clearer than the first. It is first-order and the code can mostly be understood without needing to know the definition of `mark_type_node`.

Comment (garrigue) [reply typing/btype.mli]: I have finished translating the code to the `if mark_type_node ty then ...` style.
This amounts mainly to going back to the status quo ante.
The code is in the branch https://github.com/COCTI/ocaml/tree/private_type_expr_1st_order .

The main drawback is that one has to be careful with guards, as they must be executed between the check on the level and the change of the level. So in that case the abstraction provided by your `mark_type_node` does not work.
I actually ended up introducing several bugs by forgetting to do the check.
One could of course provide a `markable_type_node` function, but this would just hide one comparison...

If you prefer the code in that branch, I'm ready to go with that.

Comment (lpw25) [reply typing/btype.mli]: The cases without guards are much clearer. If you don't mind one last pass over the code, I think that using an interface like:
```ocaml
val is_marked_type_node : type_expr -> bool
val mark_type_node : type_expr -> unit
val try_mark_type_node : type_expr -> bool
```
where `try_mark_type_node` is equivalent to `mark_type_node` in that branch, would allow you to make the cases with guards a little clearer. I would be happy to approve the version with those changes.

Comment (lpw25) [reply typing/btype.mli]: (or at least I would be happy with the parts around `mark_type_node`. I would still like to see some improvement to the code using `Internal`)

Comment (garrigue) [reply typing/btype.mli]: For `Internal`, we are working on a branch `private_type_expr_setters`.
Will merge into this PR soon.

Comment (t6s) [reply typing/btype.mli]: > The cases without guards are much clearer. If you don't mind one last pass over the code, I think that using an interface like:
> 
> ```ocaml
> val is_marked_type_node : type_expr -> bool
> val mark_type_node : type_expr -> unit
> val try_mark_type_node : type_expr -> bool
> ```
> 
> where `try_mark_type_node` is equivalent to `mark_type_node` in that branch, would allow you to make the cases with guards a little clearer. I would be happy to approve the version with those changes.

Pushed a new commit to introduce the above interface, with some name changes:
```ocaml
val not_marked_node : type_expr -> bool
val flip_mark_node : type_expr -> unit
val try_mark_node : type_expr -> bool
```

File: typing/ctype.ml
Comment (lpw25) [inline typing/ctype.ml:3270]: I think this line is a bit subtle since it deliberately includes marked nodes alongside nodes that have a level less than `level`. It would probably be clearer if this code did `if not_marked_node ty ...` and `flip_mark_node` instead of using `try_mark_node`.

**Review by lpw25 — COMMENTED** [review]
I've reviewed the latest changes. They look good.

File: typing/ctype.ml
Comment (garrigue) [reply typing/ctype.ml:3270]: Actually, this code mimics the original version, which only compared with `level`.
The current version is more robust, since we use `try_mark_node` before recursing.
The only assumption in the code is that marking a node will not set its level to `generic_level - 1`, which seems reasonable to me.
Note that separately I have removed `mirror_level` and `pivot_level` from the interface of `Btype`, which led to rewrite `Ctype.find_lowest_level` using `not_marked_node/flip_mark_node`.

**Comment by t6s** [comment]
The last commit replaces `Internal` module by `Private_type_expr` module, which provides a more modular interface with setters.
The name is also changed since `Internal` was sounding too generic.

**Comment by t6s** [comment]
Btw, I am seeing github's complaints about conflicts while the branch builds successfully at my machine. Why?

**Comment by trefis** [comment]
The conflicts that github is talking about is those you'd encounter if you were to rebase on a recent trunk.
Which you will want to do before we can merge your PR.

**Review by lpw25 — APPROVED** [review]
Latest changes look good. I think this is good to merge now.

File: typing/btype.ml
Comment (gasche) [reply typing/btype.ml]: @garrigue, I think it would be useful to answer this question (ideally in a comment in the source) or remove the line. Old comments lying around for no clear reason, that maybe we want to keep or maybe not can, turn the source into a minefield for newcomers.

**Review by gasche — COMMENTED** [review]
Looks nice, thanks!

**Comment by garrigue** [comment]
Thanks for you reviews.
I removed the commented out dead code, and rebased.
Waiting for the CI.

File: typing/btype.ml
Comment (garrigue) [reply typing/btype.ml]: Removed the comments. This was dead code since @trefis introduced scopes.

File: typing/ctype.ml
Comment (trefis) [inline typing/ctype.ml:730]: I'm a bit late to the party, but I think that comment is incorrect (and also, vaguely useless?).

File: typing/ctype.ml
Comment (trefis) [inline typing/ctype.ml:733]: Also, I think this comment (apart from being obscure to me) is at odds with the comment above the function.
So, which is it?

**Pull request merged**
